src/Npgsql/TypeHandlers/NumericHandlers/NumericHandler.cs

```diff
@@ -21,42 +21,125 @@
 // TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
 #endregion
 
-using System;
-using System.Globalization;
+using JetBrains.Annotations;
 using Npgsql.BackendMessages;
+using Npgsql.TypeHandling;
+using Npgsql.TypeMapping;
 using NpgsqlTypes;
+using System;
 using System.Data;
 using System.Diagnostics;
-using JetBrains.Annotations;
-using Npgsql.PostgresTypes;
-using Npgsql.TypeHandling;
-using Npgsql.TypeMapping;
+using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace Npgsql.TypeHandlers.NumericHandlers
 {
     [StructLayout(LayoutKind.Explicit)]
-    ref struct DecimalRaw
+    struct DecimalRaw
     {
-        internal const int SignMask = unchecked((int)0x80000000);
-        internal const int ScaleMask = 0x00FF0000;
-        internal const int ScaleShift = 16;
+        const int SignMask = unchecked((int)0x80000000);
+        const int ScaleMask = 0x00FF0000;
+        const int ScaleShift = 16;
 
+        // Do not change the order in which these fields are declared. It
+        // should be same as in the System.Decimal struct.
         [FieldOffset(0)]
-        internal decimal Value;
-
-        /// <summary>
-        /// The order of these fields is important as defined
-        /// here: https://github.com/dotnet/coreclr/blob/85374ceaed177f71472cc4c23c69daf7402e5048/src/System.Private.CoreLib/src/System/Decimal.cs#L135-L141
-        /// </summary>
-        [FieldOffset(0)]
-        internal int Flags;
+        int _flags;
         [FieldOffset(4)]
-        internal int Hi;
+        uint _high;
         [FieldOffset(8)]
-        internal int Lo;
+        uint _low;
         [FieldOffset(12)]
-        internal int Mid;
+        uint _mid;
+
+        public bool Negative => (_flags & SignMask) != 0;
+
+        public int Scale
+        {
+            get => (_flags & ScaleMask) >> ScaleShift;
+            set => _flags = (_flags & SignMask) | ((value << ScaleShift) & ScaleMask);
+        }
+
+        public uint High => _high;
+        public uint Mid => _mid;
+        public uint Low => _low;
+
+        public static void Negate(ref DecimalRaw value)
+            => value._flags ^= SignMask;
+
+        public static void Add(ref DecimalRaw value, uint addend)
```

Comment:
Any reason not to make this a regular, non-static method (i.e. `Add(uint addend)` which modifies this)? Same for the methods below.

Reply:
You are right here because the struct is used only for local variables, but as you know if it is used as parameter, then an instance method is called on a copy. Therefore, I prefer to follow the guidelines and do not implement instance methods in a mutable struct.

Reply:
Yeah, I know about "mutable structs are evil", but I think sometimes it goes a little too far - as long as everyone is aware it's not that bad. Also as long as the struct itself is passed by ref (as should be the case for big structs) it should be fine. But anyway as you wish.

Reply:
It's not a `ref struct` anymore.

Reply:
I didn't mean that it's a `ref struct`, only that it gets passed around by reference (via `ref` parameters). With a large struct, passing by `ref` reduces the copying and also makes it less error-prone for the class to be mutable...